### PR TITLE
Implement: multiple save and load for mlflow registry

### PR DESCRIPTION
**Pull request by yleilawang**
1. Implemented save_multiple and load_multiple for mlflow registry
2. Test cases for implementation.

### CODE REVIEW

File: numalogic/registry/mlflow_registry.py [hunk 1]
```diff
@@ -187,6 +189,39 @@ def load(
...
+            dict_artifacts = loaded_model.artifact.unwrap_python_model().dict_artifacts
+            artifacts_dict = {}
+            for artifact in dict_artifacts.values():
+                artifact_data = ArtifactData(
+                    artifact=artifact.artifact, metadata=metadata, extras=None
+                )
+                dynamic_key = ":".join(artifact.dkeys)
+                artifacts_dict[dynamic_key] = artifact_data
+        else:
+            artifacts_dict = None
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:222]: Can we just return the ArtifactKey instead?

File: numalogic/registry/mlflow_registry.py [hunk 2]
```diff
@@ -238,6 +276,37 @@ def save(
...
+            skeys: static key fields as list/tuple of strings
+            dict_artifacts: dict of artifacts to save
+            metadata: additional metadata surrounding the artifact that needs to be saved.
+
+        Returns
+        -------
+            mlflow ModelVersion instance
+        """
+        multiple_artifacts = CompositeModels(skeys=skeys, dict_artifacts=dict_artifacts, **metadata)
+        dkeys_list = multiple_artifacts.get_dkeys_list()
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:300]: make this function private?

File: numalogic/registry/mlflow_registry.py [hunk 3]
```diff
@@ -338,3 +407,63 @@ def __load_artifacts(
...
+
+    Attributes
+    ----------
+        skeys (KEYS): The static keys of the artifacts.
+        dict_artifacts (dict[str, KeyedArtifact]): A dictionary mapping dynamic keys to
+            `KeyedArtifact` objects.
+        metadata (META_VT): Additional metadata associated with the artifacts.
+    """
+
+    def __init__(self, skeys: KEYS, dict_artifacts: dict[str, KeyedArtifact], **metadata: META_VT):
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:452]: add __slots__ for class

File: numalogic/registry/mlflow_registry.py [hunk 4]
```diff
@@ -187,6 +189,39 @@ def load(
...
+        Returns
+        -------
+            Optional[dict[str, ArtifactData]]: A dictionary mapping joined dynamic keys
+            to the loaded artifacts, or None if no artifacts were found.
+        """
+        dkeys = self.__get_sorted_unique_dkeys(dkeys_list)
+        loaded_model = self.load(skeys=skeys, dkeys=dkeys, artifact_type="pyfunc")
+        if loaded_model is not None:
+            metadata = loaded_model.artifact.unwrap_python_model().metadata
+            dict_artifacts = loaded_model.artifact.unwrap_python_model().dict_artifacts
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:213]: can we add excption handling for this unwrapping

**Review by s0nicboOm — CHANGES_REQUESTED** [review]
Also, update the version for pyproject.toml

**Comment by ab93** [comment]
> Also, update the version for pyproject.toml

@s0nicboOm we have not released any previous versions right?

**Comment by yleilawang** [comment]
Added changes for:
1. Changed return type of load_multiple to ArtifactData, the same return type of load.
2. Slots for CompositeModels class.
3. Handling exceptions caused by unwarpping pyfunc models.
4. Private function for getting unique sorted dkey list.
5. Added test cases for cache loading for pyfunc models.

**Comment by yleilawang** [comment]
**Profiling**

`def f():
    reg.load(skeys=skeys, dkeys= ["ae", "pipeline"], artifact_type='pytorch')
    reg.load(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact_type='sklearn')
    reg.load(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact_type='sklearn')

def g():
    reg.load_multiple(skeys=skeys, dkeys_list=[["ae", "pipeline"], ["scaler", "pipeline"], ["threshold", "pipeline"]])

import timeit
a = timeit.timeit(f, number=10)
b = timeit.timeit(g, number=10)

print(a, b)`

1.0356064579991653 0.6088340830010566

File: numalogic/registry/mlflow_registry.py [hunk 5]
```diff
@@ -338,3 +406,65 @@ def __load_artifacts(
...
+
+        Returns
+        -------
+            list[list[str]]: A list of all dynamic keys in the stored artifacts.
+        """
+        dkeys_list = []
+        artifacts = self.dict_artifacts.values()
+        for artifact in artifacts:
+            dkeys_list.append(artifact.dkeys)
+        return dkeys_list
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:470]:
```suggestion
        return [ka.dkeys for _, ka  in self.dict_artifacts]
```

File: numalogic/registry/mlflow_registry.py [hunk 6]
```diff
@@ -338,3 +406,65 @@ def __load_artifacts(
...
+
+        Returns
+        -------
+        List[str]
+            A list of all unique dkeys in the stored artifacts, sorted in ascending order.
+        """
+        return list(SortedSet([dkey for dkeys in dkeys_list for dkey in dkeys]))
+
+
+class CompositeModels(mlflow.pyfunc.PythonModel):
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:426]:
```suggestion
class CompositeModel(mlflow.pyfunc.PythonModel):
```

File: numalogic/registry/mlflow_registry.py (hunk 5 above)
Comment (yleilawang) [reply numalogic/registry/mlflow_registry.py:470]: Method not needed.

**Comment by yleilawang** [comment]
**Profiling Updates**

```
reg.save(skeys=skeys, dkeys= ["ae", "pipeline"], artifact=VanillaAE(10), artifact_type='pytorch', **{"a": "b"})
reg.save(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact=StandardScaler(), artifact_type='sklearn', **{"a": "b"})
reg.save(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact=StdDevThreshold(), artifact_type='sklearn', **{"a": "b"})

output = reg.save_multiple(
    skeys=skeys, dkeys=dkeys, dict_artifacts=dict_artifacts, **{"a": "b"}
)

def f():
    reg.load(skeys=skeys, dkeys= ["ae", "pipeline"], artifact_type='pytorch')
    reg.load(skeys=skeys, dkeys= ["scaler", "pipeline"], artifact_type='sklearn')
    reg.load(skeys=skeys, dkeys= ["threshold", "pipeline"], artifact_type='sklearn')

def g():
    reg.load_multiple(skeys=skeys, dkeys=dkeys)

import timeit

a = timeit.timeit(f, number=100)
b = timeit.timeit(g, number=100)

print(a, b)
```
--> 9.074686333002319 4.829918750001525

File: numalogic/registry/mlflow_registry.py [hunk 7]
```diff
@@ -187,6 +188,43 @@ def load(
...
+            return None
+        except Exception:
+            _LOGGER.exception("Unexpected error occurred while unwrapping python model.")
+            return None
+        else:
+            return ArtifactData(
+                artifact=unwrapped_composite_model.dict_artifacts,
+                metadata=loaded_model.metadata,
+                extras=loaded_model.extras,
+            )
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:226]:
```suggestion
        return ArtifactData(
            artifact=unwrapped_composite_model.dict_artifacts,
            metadata=loaded_model.metadata,
             extras=loaded_model.extras,
        )
```

File: numalogic/registry/mlflow_registry.py [hunk 8]
```diff
@@ -238,6 +279,40 @@ def save(
...
+            dict_artifacts (dict[str, artifact_t]): Dictionary of artifacts to save.
+            **metadata (META_VT): Additional metadata to be saved with the artifacts.
+
+        Returns
+        -------
+            Optional[ModelVersion]: An instance of the MLflow ModelVersion.
+
+        """
+        if len(dict_artifacts) == 1:
+            _LOGGER.warning("Only one element in dict_artifacts. Please use save directly.")
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:306]:
```suggestion
            _LOGGER.warning("Only one element in dict_artifacts. Saving directly is recommended.")
```

File: numalogic/registry/mlflow_registry.py [hunk 9]
```diff
@@ -338,3 +413,45 @@ def __load_artifacts(
...
+    """
+
+    __slots__ = ("skeys", "dict_artifacts", "metadata")
+
+    def __init__(self, skeys: KEYS, dict_artifacts: dict[str, artifact_t], **metadata: META_VT):
+        self.skeys = skeys
+        self.dict_artifacts = dict_artifacts
+        self.metadata = metadata
+
+    def predict(self):
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:450]: Can you match the signature of the base class here? I think it takes in some parameters

**Review by ab93 — APPROVED** [review]
Looks good!

File: numalogic/registry/mlflow_registry.py [hunk 10]
```diff
@@ -18,6 +18,7 @@
 import mlflow.pyfunc
 import mlflow.pytorch
 import mlflow.sklearn
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:20]: I think we don't need these imports?

File: numalogic/registry/mlflow_registry.py [hunk 11]
```diff
@@ -238,6 +279,40 @@ def save(
...
+            dict_artifacts (dict[str, artifact_t]): Dictionary of artifacts to save.
+            **metadata (META_VT): Additional metadata to be saved with the artifacts.
+
+        Returns
+        -------
+            Optional[ModelVersion]: An instance of the MLflow ModelVersion.
+
+        """
+        if len(dict_artifacts) == 1:
+            _LOGGER.warning("Only one element in dict_artifacts. Saving directly is recommended.")
```
Comment (s0nicboOm) [inline numalogic/registry/mlflow_registry.py:306]: `_LOGGER.warning("Only one artifact present in dict_artifacts. Saving directly is recommended.")`

**Review by s0nicboOm — APPROVED** [review]
Minor comments. 
Otherwise LGTM!

File: numalogic/registry/mlflow_registry.py (hunk 10 above)
Comment (yleilawang) [reply numalogic/registry/mlflow_registry.py:20]: Test cases still need these imports. Otherwise will cause connection errors in tests.

Comment (ab93) [reply numalogic/registry/mlflow_registry.py:20]: Agreed. Let's optimize the imports

Comment (ab93) [reply numalogic/registry/mlflow_registry.py:20]: @yleilawang a good rule of thumb is to make the test cases to work around the actual code. Not the other way round! :)

Comment (yleilawang) [reply numalogic/registry/mlflow_registry.py:20]: Can we move these imports to test file? Just tried and it didn't cause errors. @ab93 @s0nicboOm

Comment (ab93) [reply numalogic/registry/mlflow_registry.py:20]: Yes test file sounds good!

**Pull request merged**